{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "42f075d6",
   "metadata": {},
   "source": [
    "# Extract cell-type-specific embeddings from fine-tuned ChromBERT.\n",
    "\n",
    "All commands below fine-tune ChromBERT on cell-type-specific accessibility data (BigWig + peaks) \n",
    "and then extract embeddings from the fine-tuned model. If a fine-tuned checkpoint is provided, \n",
    "fine-tuning is skipped and embeddings are generated directly from the checkpoint.\n",
    "\n",
    "- ``embed_cell_cistrome``: extracts cistrome embeddings\n",
    "- ``embed_cell_gene``: extracts gene embeddings reflecting cell-type-specific regulatory context\n",
    "- ``embed_cell_region``: extracts region embeddings reflecting cell-type-specific regulatory patterns\n",
    "- ``embed_cell_regulator``: extracts regulator embeddings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d581761",
   "metadata": {},
   "source": [
    "# Example: \n",
    "Extract cell-type-specific embeddings for myoblast, including cistrome, gene, region, and regulator embeddings.\n",
    "\n",
    "``Recommended``: First run infer_cell_trn to generate a fine-tuned checkpoint and infer key regulators/TRNs. Then use that checkpoint to extract cell-type-specific embeddings (cistrome, gene, region, and regulator)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "49383189",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir(\"/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli\")\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"]='0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19d8a21a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import h5py\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "015146ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_file = '../data/CTCF_ENCFF664UGR_sample100.bed' # only for example\n",
    "ft_ckpt_dir = \"/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/**/*.ckpt\" # if you have already run infer_cell_trn, you can use the fine-tuned checkpoint to extract cell-type-specific embeddings\n",
    "\n",
    "ft_ckpt = glob.glob(ft_ckpt_dir, recursive=True)[0]\n",
    "ft_ckpt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7e616d9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools [OPTIONS] COMMAND [ARGS]...\n",
      "\n",
      "  Type -h or --help after any subcommand for more information.\n",
      "\n",
      "Options:\n",
      "  -v, --verbose  Verbose logging\n",
      "  -d, --debug    Post mortem debugging\n",
      "  -V, --version  Show the version and exit.\n",
      "  -h, --help     Show this message and exit.\n",
      "\n",
      "Commands:\n",
      "  embed_cell_cistrome         Extract cell-specific cistrome embeddings...\n",
      "  embed_cell_gene             Extract cell-specific gene embeddings\n",
      "  embed_cell_region           Extract cell-specific region embeddings\n",
      "  embed_cell_regulator        Extract cell-specific regulator embeddings...\n",
      "  embed_cistrome              Extract general cistrome embeddings on...\n",
      "  embed_gene                  Extract general gene embeddings\n",
      "  embed_region                Extract general region embeddings\n",
      "  embed_regulator             Extract general regulator embeddings on...\n",
      "  find_driver_in_dual_region  Find driver factors in dual functional...\n",
      "  find_driver_in_transition   Find driver factors in cell state transitions.\n",
      "  impute_cistrome             Impute cistromes on specified regions\n",
      "  infer_cell_trn              Infer cell-specific TRN (Transition...\n",
      "  infer_trn                   Infer general TRN (Transition Regulatory...\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29450674",
   "metadata": {},
   "source": [
    "## Embed cistromes\n",
    "Extract cell-specific cistrome embeddings on specified regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "688d7d78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools embed_cell_cistrome [OPTIONS]\n",
      "\n",
      "  Extract cell-specific cistrome embeddings on specified regions\n",
      "\n",
      "Options:\n",
      "  --region FILE                   Region file where cistrome embeddings will\n",
      "                                  be computed.  [required]\n",
      "  --cistrome TEXT                 GSM/ENCODE id or factor:cell, e.g. ENCSR...\n",
      "                                  or GSM... or ATAC-seq:HEK293T or\n",
      "                                  BCL11A:GM12878. Use ';' to separate\n",
      "                                  multiple.  [required]\n",
      "  --cell-type-bw FILE             Cell type accessibility BigWig file.\n",
      "                                  Required if --ft-ckpt is not provided.\n",
      "  --cell-type-peak FILE           Cell type accessibility Peak BED file.\n",
      "                                  Required if --ft-ckpt is not provided.\n",
      "  --ft-ckpt FILE                  Fine-tuned ChromBERT checkpoint. If\n",
      "                                  provided, skip fine-tuning and use this\n",
      "                                  ckpt. If not provided, you must provide\n",
      "                                  --cell-type-bw and --cell-type-peak to train\n",
      "                                  a cell-specific model.\n",
      "  --odir DIRECTORY                Output directory.  [default: ./output]\n",
      "  --genome [hg38|mm10]            Genome.  [default: hg38]\n",
      "  --resolution [1kb|200bp|2kb|4kb]\n",
      "                                  Resolution. Mouse only supports 1kb\n",
      "                                  resolution.  [default: 1kb]\n",
      "  --mode [fast|full]              Fast: downsample regions to 20k for\n",
      "                                  training; Full: use all regions.  [default:\n",
      "                                  fast]\n",
      "  --batch-size INTEGER            Batch size.  [default: 4]\n",
      "  --num-workers INTEGER           Dataloader workers.  [default: 8]\n",
      "  --chrombert-cache-dir DIRECTORY\n",
      "                                  ChromBERT cache dir (contains config/\n",
      "                                  checkpoint/ etc).  [default:\n",
      "                                  ~/.cache/chrombert/data]\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools embed_cell_cistrome -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "330f53e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # if you not provided checkpoint\n",
    "# !chrombert-tools embed_cell_cistrome \\\n",
    "#     --cell-type-bw \"../data/myoblast_ENCFF149ERN_signal.bigwig\" \\\n",
    "#     --cell-type-peak \"../data/myoblast_ENCFF647RNC_peak.bed\" \\\n",
    "#     --region {region_file} \\\n",
    "#     --cistrome \"GSM837613;CTCF:HSMM;h3k27ac:LHCNM2\" \\\n",
    "#     --odir \"./output_cell_specific_emb_cistrome\" \\\n",
    "#     --genome \"hg38\" \\\n",
    "#     --resolution \"1kb\" 2> \"./tmp/infer_cell_trn.stderr.log\" # redirect stderr to log file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b9913143",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage 1: Preparing regions\n",
      "Region summary - total: 100, overlapping with ChromBERT: 100 (one region may overlap multiple ChromBERT regions), non-overlapping: 0\n",
      "Note: All cistrome names were converted to lowercase for matching.\n",
      "Cistromes count summary - requested: 3, matched in ChromBERT meta: 3, not found: 0, not found cistromes: []\n",
      "ChromBERT cistromes metas: /mnt/Storage/home/chenqianqian/.cache/chrombert/data/config/hg38_6k_meta.tsv\n",
      "Found 3 cistromes: ['gsm837613', 'ctcf:hsmm', 'h3k27ac:lhcnm2']\n",
      "Finished stage 1\n",
      "Stage 2: Using provided fine-tuned ChromBERT checkpoint: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 110/110 parameters\n",
      "Finished stage 2\n",
      "Stage 3: Computing cell-specific cistrome embeddings\n",
      "Your supervised_file does not contain the 'label' column. Please verify whether ground truth column ('label') is required. If it is not needed, you may disregard this message.\n",
      "Your supervised_file does not contain the 'label' column. Please verify whether ground truth column ('label') is required. If it is not needed, you may disregard this message.\n",
      "Computing cistrome embeddings: 100%|████████████| 25/25 [00:05<00:00,  4.17it/s]\n",
      "Finished stage 3\n",
      "\n",
      "Finished all stages!\n",
      "Cell-specific mean cistrome embeddings saved to: ./output_cell_specific_emb_cistrome/cell_specific_mean_cistrome_emb.pkl\n",
      "Cell-specific cistrome embeddings on regions saved to: ./output_cell_specific_emb_cistrome/cell_specific_cistrome_emb_on_region.hdf5\n",
      "Number of cistromes embedded: 3\n"
     ]
    }
   ],
   "source": [
    "# region_file is a bed file\n",
    "!chrombert-tools embed_cell_cistrome \\\n",
    "    --region {region_file} \\\n",
    "    --ft-ckpt {ft_ckpt} \\\n",
    "    --cistrome \"GSM837613;CTCF:HSMM;h3k27ac:LHCNM2\" \\\n",
    "    --odir \"./output_cell_specific_emb_cistrome\" \\\n",
    "    --genome \"hg38\" \\\n",
    "    --resolution \"1kb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b9640913",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cell_specific_mean_cistrome_emb.pkl: one 768-dim vector per cistrome (averaged across all regions)\n",
    "with open(\"./output_cell_specific_emb_cistrome/cell_specific_mean_cistrome_emb.pkl\", \"rb\") as f:\n",
    "    mean_cistrome_emb_dict2 = pickle.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0a000d3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsm837613 (768,)\n",
      "ctcf:hsmm (768,)\n",
      "h3k27ac:lhcnm2 (768,)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for key, value in mean_cistrome_emb_dict2.items():\n",
    "    print(key, value.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "26be325d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<KeysViewHDF5 ['emb', 'region']>\n",
      "<KeysViewHDF5 ['ctcf:hsmm', 'gsm837613', 'h3k27ac:lhcnm2']>\n",
      "ctcf:hsmm (100, 768)\n",
      "gsm837613 (100, 768)\n",
      "h3k27ac:lhcnm2 (100, 768)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[       1, 37989946, 37990368,    32658],\n",
       "       [      11,  2400199,  2400617,   289179],\n",
       "       [      12,  6778809,  6779319,   391108],\n",
       "       [      12, 52980788, 52981316,   424926],\n",
       "       [      12, 53676021, 53676448,   425578],\n",
       "       [      14, 21092401, 21092968,   560876],\n",
       "       [      14, 23057979, 23058458,   562483],\n",
       "       [      14, 23120727, 23121190,   562542],\n",
       "       [      14, 23379895, 23380314,   562781],\n",
       "       [      14, 23588973, 23589439,   562958]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cell_specific_cistrome_emb_on_region.hdf5: per-region embeddings; each cistrome has a matrix of shape (N_regions, 768)\n",
    "# region: [chrom, start, end, build_region_index]\n",
    "# build_region_index maps to ChromBERT's reference regions\n",
    "with h5py.File(\"./output_cell_specific_emb_cistrome/cell_specific_cistrome_emb_on_region.hdf5\", \"r\") as f:\n",
    "    print(f.keys())\n",
    "    print(f['emb'].keys())\n",
    "    emb_key = list(f['emb'].keys())\n",
    "    region = f['region'][:]\n",
    "    for key in emb_key:\n",
    "        emb = f['emb'][key][:]\n",
    "        print(key, emb.shape)\n",
    "\n",
    "region[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "837860cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>build_region_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>37989946</td>\n",
       "      <td>37990368</td>\n",
       "      <td>32658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2400199</td>\n",
       "      <td>2400617</td>\n",
       "      <td>289179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>6778809</td>\n",
       "      <td>6779319</td>\n",
       "      <td>391108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>52980788</td>\n",
       "      <td>52981316</td>\n",
       "      <td>424926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12</td>\n",
       "      <td>53676021</td>\n",
       "      <td>53676448</td>\n",
       "      <td>425578</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     start       end  build_region_index\n",
       "0   chr1  37989946  37990368               32658\n",
       "1  chr11   2400199   2400617              289179\n",
       "2  chr12   6778809   6779319              391108\n",
       "3  chr12  52980788  52981316              424926\n",
       "4  chr12  53676021  53676448              425578"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# overlap_region.bed: input regions overlapped with ChromBERT's reference regions; contains columns: chrom, start, end, build_region_index\n",
    "overlap_region = pd.read_csv(\"./output_cell_specific_emb_cistrome/overlap_region.bed\",sep='\\t',header=None, names=['chrom','start','end','build_region_index'])\n",
    "overlap_region.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44907674",
   "metadata": {},
   "source": [
    "## Embed genes\n",
    "Extract cell-specific gene embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "131e2902",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools embed_cell_gene [OPTIONS]\n",
      "\n",
      "  Extract cell-specific gene embeddings\n",
      "\n",
      "Options:\n",
      "  --gene TEXT                     Gene symbols or IDs. e.g.\n",
      "                                  ENSG00000170921;TANC2;DPYD. Use ';' to\n",
      "                                  separate multiple genes.  [required]\n",
      "  --cell-type-bw FILE             Cell type accessibility BigWig file.\n",
      "  --cell-type-peak FILE           Cell type accessibility Peak BED file.\n",
      "  --ft-ckpt FILE                  Fine-tuned ChromBERT checkpoint. If\n",
      "                                  provided, skip fine-tuning and use this\n",
      "                                  ckpt. If you not provide, you should provide\n",
      "                                  --cell-type-bw and --cell-type-peak to train\n",
      "                                  a cell-specific model.\n",
      "  --odir DIRECTORY                Output directory.  [default: ./output]\n",
      "  --genome [hg38|mm10]            Genome.  [default: hg38]\n",
      "  --resolution [1kb|200bp|2kb|4kb]\n",
      "                                  Resolution. Mouse only supports 1kb\n",
      "                                  resolution.  [default: 1kb]\n",
      "  --mode [fast|full]              Fast: downsample regions to 20k for\n",
      "                                  training; Full: use all regions.  [default:\n",
      "                                  fast]\n",
      "  --batch-size INTEGER            Batch size.  [default: 4]\n",
      "  --chrombert-cache-dir DIRECTORY\n",
      "                                  ChromBERT cache dir. If you use\n",
      "                                  `chrombert_prepare_env`, you usually don't\n",
      "                                  need to provide this.  [default:\n",
      "                                  ~/.cache/chrombert/data]\n",
      "  --chrombert-region-file FILE    ChromBERT region BED file. If not provided,\n",
      "                                  use the default\n",
      "                                  {genome}_{nd}_{resolution}_region.bed in the\n",
      "                                  cache dir.\n",
      "  --chrombert-gene-meta FILE      ChromBERT gene meta TSV. If not provided,\n",
      "                                  try {genome}_{resolution}_gene_meta.tsv in\n",
      "                                  the cache dir.\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools embed_cell_gene -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "eada9fa5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished stage 1\n",
      "Stage 2: Using provided fine-tuned ChromBERT checkpoint: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 110/110 parameters\n",
      "Finished stage 2\n",
      "Stage 3: Computing cell-specific gene embeddings\n",
      "Your supervised_file does not contain the 'label' column. Please verify whether ground truth column ('label') is required. If it is not needed, you may disregard this message.\n",
      "Computing region embeddings: 100%|████████████████| 2/2 [00:01<00:00,  1.46it/s]\n",
      "Finished stage 3\n",
      "\n",
      "Finished all stages!\n",
      "Note: All gene names were converted to lowercase for matching.\n",
      "Gene count summary - requested: 5, matched: 5, not found: 0, not found gene names: []\n",
      "Cell-specific ChromBERT model saved to: None\n",
      "Cell-specific gene embeddings saved to: ./output_emb_cell_genes/cell_specific_gene_embs_dict.pkl\n",
      "Matched gene meta saved to: ./output_emb_cell_genes/overlap_genes_meta.tsv\n"
     ]
    }
   ],
   "source": [
    "! chrombert-tools embed_cell_gene \\\n",
    "    --gene \"myod1;myf5;tp53;brd4;ENSG00000009709\" \\\n",
    "    --odir \"./output_emb_cell_genes\" \\\n",
    "    --ft-ckpt {ft_ckpt} \\\n",
    "    --genome \"hg38\" \\\n",
    "    --resolution \"1kb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "080c0785",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ensg00000170921 (768,)\n",
      "tanc2 (768,)\n",
      "ensg00000200997 (768,)\n",
      "dpyd (768,)\n",
      "snora70 (768,)\n",
      "tp53 (768,)\n",
      "brd4 (768,)\n"
     ]
    }
   ],
   "source": [
    "# embs_dict.pkl: gene embed dict\n",
    "with open(\"./output_emb_genes/embs_dict.pkl\", \"rb\") as f:\n",
    "    gene_emb_dict = pickle.load(f)\n",
    "for key, value in gene_emb_dict.items():\n",
    "    print(key, value.shape)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc1c6532",
   "metadata": {},
   "source": [
    "## Embed region\n",
    "Extract cell-specific region embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "75dbe09d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools embed_cell_region [OPTIONS]\n",
      "\n",
      "  Extract cell-specific region embeddings\n",
      "\n",
      "Options:\n",
      "  --region FILE                   Region file to compute embeddings for.\n",
      "                                  [required]\n",
      "  --cell-type-bw FILE             Cell type accessibility BigWig file.\n",
      "                                  Required if --ft-ckpt is not provided.\n",
      "  --cell-type-peak FILE           Cell type accessibility Peak BED file.\n",
      "                                  Required if --ft-ckpt is not provided.\n",
      "  --ft-ckpt FILE                  Fine-tuned ChromBERT checkpoint. If\n",
      "                                  provided, skip fine-tuning and use this\n",
      "                                  ckpt. If not provided, you must provide\n",
      "                                  --cell-type-bw and --cell-type-peak to train\n",
      "                                  a cell-specific model.\n",
      "  --odir DIRECTORY                Output directory.  [default: ./output]\n",
      "  --genome [hg38|mm10]            Genome.  [default: hg38]\n",
      "  --resolution [1kb|200bp|2kb|4kb]\n",
      "                                  Resolution. Mouse only supports 1kb\n",
      "                                  resolution.  [default: 1kb]\n",
      "  --mode [fast|full]              Fast: downsample regions to 20k for\n",
      "                                  training; Full: use all regions.  [default:\n",
      "                                  fast]\n",
      "  --batch-size INTEGER            Batch size.  [default: 4]\n",
      "  --chrombert-cache-dir DIRECTORY\n",
      "                                  ChromBERT cache dir. If you use\n",
      "                                  `chrombert_prepare_env`, you usually don't\n",
      "                                  need to provide this.  [default: /mnt/Storag\n",
      "                                  e/home/chenqianqian/.cache/chrombert/data]\n",
      "  --chrombert-region-file FILE    ChromBERT region BED file. If not provided,\n",
      "                                  use the default\n",
      "                                  {genome}_{nd}_{resolution}_region.bed in the\n",
      "                                  cache dir.\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools embed_cell_region -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "37970026",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage 1: Overlapping focus regions with ChromBERT regions\n",
      "Region summary - total: 100, overlapping with ChromBERT: 100 (one region may overlap multiple ChromBERT regions), non-overlapping: 0\n",
      "Found 100 overlapping regions\n",
      "Finished stage 1\n",
      "Stage 2: Using provided fine-tuned ChromBERT checkpoint: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 110/110 parameters\n",
      "Finished stage 2\n",
      "Stage 3: Computing cell-specific region embeddings\n",
      "Your supervised_file does not contain the 'label' column. Please verify whether ground truth column ('label') is required. If it is not needed, you may disregard this message.\n",
      "Computing region embeddings: 100%|██████████████| 25/25 [00:05<00:00,  4.25it/s]\n",
      "Finished stage 3\n",
      "\n",
      "Finished all stages!\n",
      "Focus region summary - total: 100, overlapping with ChromBERT: 100, non-overlapping: 0\n",
      "Note: It is possible for a single region to overlap multiple ChromBERT regions.\n",
      "Overlapping regions BED file: ./output_cell_specific_emb_region/overlap_region.bed\n",
      "Non-overlapping regions BED file: ./output_cell_specific_emb_region/no_overlap_region.bed\n",
      "Cell-specific region embeddings saved to: ./output_cell_specific_emb_region/cell_specific_overlap_region_emb.npy\n"
     ]
    }
   ],
   "source": [
    "! chrombert-tools embed_cell_region \\\n",
    "    --region {region_file} \\\n",
    "    --ft-ckpt {ft_ckpt} \\\n",
    "    --odir \"./output_cell_specific_emb_region\" \\\n",
    "    --genome \"hg38\" \\\n",
    "    --resolution \"1kb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9590aa05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(100, 768)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(100,\n",
       "    chrom     start       end  build_region_index\n",
       " 0   chr1  37989946  37990368               32658\n",
       " 1  chr11   2400199   2400617              289179\n",
       " 2  chr12   6778809   6779319              391108\n",
       " 3  chr12  52980788  52981316              424926\n",
       " 4  chr12  53676021  53676448              425578)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cell_specific_overlap_region_emb.npy: one 768-dim vector per region\n",
    "overlap_region_emb = np.load(\"./output_cell_specific_emb_region/cell_specific_overlap_region_emb.npy\")\n",
    "print(overlap_region_emb.shape)\n",
    "\n",
    "# overlap_region.bed: input regions overlapped with ChromBERT's reference regions; contains columns: chrom, start, end, build_region_index\n",
    "\n",
    "overlap_region = pd.read_csv(\"./output_emb_region/overlap_region.bed\",sep='\\t',header=None, names=['chrom','start','end','build_region_index'])\n",
    "len(overlap_region),overlap_region.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb9e6f28",
   "metadata": {},
   "source": [
    "## Embed regulator\n",
    "Extract cell-specific regulator embeddings on specified regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f5e6e0d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools embed_cell_regulator [OPTIONS]\n",
      "\n",
      "  Extract cell-specific regulator embeddings on specified regions\n",
      "\n",
      "Options:\n",
      "  --region FILE                   Region file where regulator embeddings will\n",
      "                                  be computed.  [required]\n",
      "  --regulator TEXT                Regulators of interest, e.g. EZH2 or\n",
      "                                  EZH2;BRD4. Use ';' to separate multiple\n",
      "                                  regulators.  [required]\n",
      "  --cell-type-bw FILE             Cell type accessibility BigWig file.\n",
      "                                  Required if --ft-ckpt is not provided.\n",
      "  --cell-type-peak FILE           Cell type accessibility Peak BED file.\n",
      "                                  Required if --ft-ckpt is not provided.\n",
      "  --ft-ckpt FILE                  Fine-tuned ChromBERT checkpoint. If\n",
      "                                  provided, skip fine-tuning and use this\n",
      "                                  ckpt. If not provided, you must provide\n",
      "                                  --cell-type-bw and --cell-type-peak to train\n",
      "                                  a cell-specific model.\n",
      "  --odir DIRECTORY                Output directory.  [default: ./output]\n",
      "  --genome [hg38|mm10]            Genome.  [default: hg38]\n",
      "  --resolution [1kb|200bp|2kb|4kb]\n",
      "                                  Resolution. Mouse only supports 1kb\n",
      "                                  resolution.  [default: 1kb]\n",
      "  --mode [fast|full]              Fast: downsample regions to 20k for\n",
      "                                  training; Full: use all regions.  [default:\n",
      "                                  fast]\n",
      "  --batch-size INTEGER            Batch size.  [default: 4]\n",
      "  --num-workers INTEGER           Dataloader workers.  [default: 8]\n",
      "  --chrombert-cache-dir DIRECTORY\n",
      "                                  ChromBERT cache dir (contains config/\n",
      "                                  checkpoint/ etc).  [default: /mnt/Storage/ho\n",
      "                                  me/chenqianqian/.cache/chrombert/data]\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools embed_cell_regulator -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "120b2942",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage 1: Preparing regions and regulators\n",
      "Region summary - total: 100, overlapping with ChromBERT: 100 (one region may overlap multiple ChromBERT regions), non-overlapping: 0\n",
      "Note: All regulator names were converted to lowercase for matching.\n",
      "Regulator count summary - requested: 6, matched in ChromBERT: 5, not found: 1, not found regulator: ['foxa3']\n",
      "ChromBERT regulators: /mnt/Storage/home/chenqianqian/.cache/chrombert/data/config/hg38_6k_regulators_list.txt\n",
      "Found 5 regulators: ['myod1', 'ctcf', 'myf5', 'brd4', 'ezh2']\n",
      "Finished stage 1\n",
      "Stage 2: Using provided fine-tuned ChromBERT checkpoint: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_infer_cell_trn/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=3-step=226.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 110/110 parameters\n",
      "Finished stage 2\n",
      "Stage 3: Computing cell-specific regulator embeddings\n",
      "Your supervised_file does not contain the 'label' column. Please verify whether ground truth column ('label') is required. If it is not needed, you may disregard this message.\n",
      "Your supervised_file does not contain the 'label' column. Please verify whether ground truth column ('label') is required. If it is not needed, you may disregard this message.\n",
      "Computing regulator embeddings: 100%|███████████| 25/25 [00:05<00:00,  4.20it/s]\n",
      "Finished stage 3\n",
      "\n",
      "Finished all stages!\n",
      "Cell-specific mean regulator embeddings saved to: ./output_cell_specific_emb_regulator/cell_specific_mean_regulator_emb.pkl\n",
      "Cell-specific regulator embeddings on regions saved to: ./output_cell_specific_emb_regulator/cell_specific_regulator_emb_on_region.hdf5\n",
      "Number of regulators embedded: 5\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools embed_cell_regulator \\\n",
    "    --region {region_file} \\\n",
    "    --ft-ckpt {ft_ckpt} \\\n",
    "    --regulator \"EZH2;BRD4;CTCF;FOXA3;myod1;myF5\" \\\n",
    "    --odir \"./output_cell_specific_emb_regulator\" \\\n",
    "    --genome \"hg38\" \\\n",
    "    --resolution \"1kb\"\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "aeb9ba7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "myod1 (768,)\n",
      "ctcf (768,)\n",
      "myf5 (768,)\n",
      "brd4 (768,)\n",
      "ezh2 (768,)\n"
     ]
    }
   ],
   "source": [
    "# cell_specific_mean_regulator_emb.pkl: one 768-dim vector per regulator (averaged across all regions)\n",
    "with open(\"./output_cell_specific_emb_regulator/cell_specific_mean_regulator_emb.pkl\", \"rb\") as f:\n",
    "    mean_regulator_emb_dict = pickle.load(f)\n",
    "\n",
    "for key, value in mean_regulator_emb_dict.items():\n",
    "    print(key, value.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "47816057",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<KeysViewHDF5 ['emb', 'region']>\n",
      "<KeysViewHDF5 ['brd4', 'ctcf', 'ezh2', 'myf5', 'myod1']>\n",
      "brd4 (100, 768)\n",
      "ctcf (100, 768)\n",
      "ezh2 (100, 768)\n",
      "myf5 (100, 768)\n",
      "myod1 (100, 768)\n"
     ]
    }
   ],
   "source": [
    "# cell_specific_regulator_emb_on_region.hdf5: Per-region embeddings: matrix of shape (N_regions, 768) for each regulator\n",
    "with h5py.File(\"./output_cell_specific_emb_regulator/cell_specific_regulator_emb_on_region.hdf5\", \"r\") as f:\n",
    "    print(f.keys())\n",
    "    print(f['emb'].keys())\n",
    "    emb_key = list(f['emb'].keys())\n",
    "    region = f['region'][:]\n",
    "    for key in emb_key:\n",
    "        emb = f['emb'][key][:]\n",
    "        print(key, emb.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1e7120a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "chrombert",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
