{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3173608b",
   "metadata": {},
   "source": [
    "# Identify driver factors in cell state transitions\n",
    "\n",
    "The ``find_driver_in_transition`` command identifies key transcription factors that drive changes in gene expression and/or chromatin accessibility during cell state transitions (e.g., differentiation or reprogramming).\n",
    "\n",
    "You can run this command with:\n",
    "- expression only,\n",
    "- accessibility only, or\n",
    "- both expression and accessibility.\n",
    "\n",
    "Provide the corresponding input files for the analyses you want to perform."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d727b92",
   "metadata": {},
   "source": [
    "**Note**: The remaining examples will only show the direct command usage. \n",
    "\n",
    "If you need to use Singularity container, please refer to the [`singularity_use.ipynb`](singularity_use.ipynb) tutorial for detailed instructions on using `singularity exec` with `chrombert-tools`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "769416ae",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "## Example:\n",
    "\n",
    "Find driver regulators in fibroblast-to-myoblast transition using both expression and accessibility\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "601800d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "workdir=\"/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli\"\n",
    "os.chdir(workdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8701d3db",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"CUDA_VISIBLE_DEVICES\"]='2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b389891c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools [OPTIONS] COMMAND [ARGS]...\n",
      "\n",
      "  Type -h or --help after any subcommand for more information.\n",
      "\n",
      "Options:\n",
      "  -v, --verbose  Verbose logging\n",
      "  -d, --debug    Post mortem debugging\n",
      "  -V, --version  Show the version and exit.\n",
      "  -h, --help     Show this message and exit.\n",
      "\n",
      "Commands:\n",
      "  embed_cell_cistrome             Extract cell-specific cistrome...\n",
      "  embed_cell_gene                 Extract cell-specific gene embeddings\n",
      "  embed_cell_region               Extract cell-specific region embeddings\n",
      "  embed_cell_regulator            Extract cell-specific regulator...\n",
      "  embed_cistrome                  Extract general cistrome embeddings on...\n",
      "  embed_gene                      Extract general gene embeddings\n",
      "  embed_region                    Extract general region embeddings\n",
      "  embed_regulator                 Extract general regulator embeddings on...\n",
      "  find_context_specific_cofactor  Find context-specific cofactors in...\n",
      "  find_driver_in_transition       Find driver factors in cell state...\n",
      "  impute_cistrome                 Impute cistromes on specified regions\n",
      "  infer_cell_regulator_network    Infer cell-specific key regulators and...\n",
      "  infer_ep                        Infer enhancer-promoter loop\n",
      "  infer_regulator_network         Infer general regulator-regulator network\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6cf7c3d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools find_driver_in_transition [OPTIONS]\n",
      "\n",
      "  Find driver factors in cell state transitions.\n",
      "\n",
      "  This tool identifies key transcription factors that drive cell state\n",
      "  transitions by analyzing changes in gene expression and/or chromatin\n",
      "  accessibility between two cell states.\n",
      "\n",
      "  You must provide at least one of the following: - Expression data (--exp-\n",
      "  tpm1 and --exp-tpm2) - Accessibility data (--acc-peak1, --acc-peak2, --acc-\n",
      "  signal1, --acc-signal2)\n",
      "\n",
      "  Providing both expression and accessibility data yields more confident\n",
      "  results.\n",
      "\n",
      "Options:\n",
      "  --exp-tpm1 FILE                 Expression (TPM) file for cell state 1. CSV\n",
      "                                  format with 'gene_id' and 'tpm' columns.\n",
      "  --exp-tpm2 FILE                 Expression (TPM) file for cell state 2. CSV\n",
      "                                  format with 'gene_id' and 'tpm' columns.\n",
      "  --acc-peak1 FILE                Chromatin accessibility peak BED file for\n",
      "                                  cell state 1.\n",
      "  --acc-peak2 FILE                Chromatin accessibility peak BED file for\n",
      "                                  cell state 2.\n",
      "  --acc-signal1 FILE              Chromatin accessibility signal BigWig file\n",
      "                                  for cell state 1.\n",
      "  --acc-signal2 FILE              Chromatin accessibility signal BigWig file\n",
      "                                  for cell state 2.\n",
      "  --direction [2-1|1-2]           Direction of cell state transition: '2-1'\n",
      "                                  means from state 1 to state 2; '1-2' means\n",
      "                                  from state 2 to state 1.  [default: 2-1]\n",
      "  --odir DIRECTORY                Output directory.  [default: ./output]\n",
      "  --genome [hg38|mm10]            Reference genome.  [default: hg38]\n",
      "  --resolution [200bp|1kb|2kb|4kb]\n",
      "                                  ChromBERT resolution.  [default: 1kb]\n",
      "  --mode [fast|normal]            Training mode: 'fast' downsamples to 20k\n",
      "                                  regions for quicker training; 'normal' uses\n",
      "                                  all regions.  [default: fast]\n",
      "  --ft-ckpt-exp FILE              Fine-tuned ChromBERT checkpoint file for\n",
      "                                  expression changes.\n",
      "  --ft-ckpt-acc FILE              Fine-tuned ChromBERT checkpoint file for\n",
      "                                  chromatin accessibility changes.\n",
      "  --chrombert-cache-dir DIRECTORY\n",
      "                                  ChromBERT cache directory (containing\n",
      "                                  config/ and anno/ subfolders).  [default:\n",
      "                                  ~/.cache/chrombert/data]\n",
      "  --batch-size INTEGER            Batch size. Increase this value if you have\n",
      "                                  sufficient GPU memory.  [default: 4]\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools find_driver_in_transition -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63fbac86",
   "metadata": {},
   "source": [
    "## Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1f6b6c63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage 1: prepare dataset\n",
      "Expression dataset already exists in output_find_driver_in_transition/exp/dataset\n",
      "Processing stage 1: prepare chromatin accessibility dataset\n",
      "Finished Stage 1\n",
      "Whether to train ChromBERT to predict expression changes in cell state transition: True\n",
      "Whether to train ChromBERT to predict chromatin accessibility changes in cell state transition: True\n",
      "Processing stage 2 (exp): train ChromBERT to predict expression changes in cell state transition\n",
      "Stage 2 (exp): train ChromBERT to predict expression changes in cell state transition\n",
      "\n",
      "[Attempt 0/2] seed=55\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Epoch 0:  20%|████▍                 | 336/1688 [03:54<15:43,  1.43it/s, v_num=0]\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/105 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 105/105 [00:45<00:00,  2.30it/s]\u001b[A\n",
      "Epoch 0:  40%|▍| 672/1688 [08:36<13:01,  1.30it/s, v_num=0, default_validation/r\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/105 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 105/105 [00:45<00:00,  2.31it/s]\u001b[A\n",
      "Epoch 0:  60%|▌| 1008/1688 [13:16<08:57,  1.26it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/105 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 105/105 [00:45<00:00,  2.31it/s]\u001b[A\n",
      "Epoch 0:  80%|▊| 1344/1688 [17:57<04:35,  1.25it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/105 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 105/105 [00:45<00:00,  2.31it/s]\u001b[A\n",
      "Epoch 0: 100%|▉| 1680/1688 [22:37<00:06,  1.24it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/105 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 105/105 [00:45<00:00,  2.31it/s]\u001b[A\n",
      "Epoch 1:  20%|▏| 336/1688 [03:50<15:29,  1.46it/s, v_num=0, default_validation/r\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/105 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 105/105 [00:45<00:00,  2.31it/s]\u001b[A\n",
      "Epoch 1:  20%|▏| 336/1688 [04:37<18:37,  1.21it/s, v_num=0, default_validation/r\u001b[A\n",
      "Evaluating the finetuned model performance\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/exp/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=0-step=21.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 110/110 parameters\n",
      "ft_ckpt: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/exp/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=0-step=21.ckpt, test_metrics: {'pearsonr': 0.2513643205165863, 'spearmanr': 0.27217140793800354, 'mse': 0.9176307320594788, 'mae': 0.6412981152534485, 'r2': -0.16728761549162963}\n",
      "Attempt metrics: pearsonr=0.2513643205165863\n",
      "Accepted run (pearsonr=0.2514 >= 0.2).\n",
      "\n",
      "Finished stage 2: obtained a fine-tuned ChromBERT\n",
      "Best pearsonr=0.2513643205165863, metrics={'pearsonr': 0.2513643205165863, 'spearmanr': 0.27217140793800354, 'mse': 0.9176307320594788, 'mae': 0.6412981152534485, 'r2': -0.16728761549162963, 'ft_ckpt': '/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/exp/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=0-step=21.ckpt'}\n",
      "Finished stage 2 (exp): train ChromBERT to predict expression changes in cell state transition\n",
      "Stage 3 (exp): infer driver factors in different expression activity genes\n",
      "                         factors  similarity  rank\n",
      "0                           cbx6    0.961502     1\n",
      "1     histone lysine acetylation    0.962091     2\n",
      "2   histone lysine crotonylation    0.962891     3\n",
      "3                           cbx7    0.963570     4\n",
      "4                           h2ax    0.968699     5\n",
      "5                        h2ak5ac    0.970900     6\n",
      "6                       h3k36me1    0.971652     7\n",
      "7                       h3k79me1    0.974935     8\n",
      "8                           cbx8    0.976251     9\n",
      "9                      h2bk120ac    0.976620    10\n",
      "10                     h4tetraac    0.976651    11\n",
      "11                      h2bk20ac    0.977353    12\n",
      "12                       h4k91ac    0.979013    13\n",
      "13                         ring1    0.979848    14\n",
      "14                      h2bk12ac    0.980100    15\n",
      "15                          brd7    0.980695    16\n",
      "16                      h2bk15ac    0.981230    17\n",
      "17                          hira    0.981533    18\n",
      "18                         klf11    0.981574    19\n",
      "19                           rb1    0.981926    20\n",
      "20                         prkdc    0.982076    21\n",
      "21                       h2bk5ac    0.982828    22\n",
      "22                        h3k4ac    0.983176    23\n",
      "23                       h3k56ac    0.983241    24\n",
      "24                     h3k9/14ac    0.983257    25\n",
      "Finished stage 4 exp: infer driver factors in cell state transition (top 25):\n",
      "   factors  similarity  rank\n",
      "0     cbx6    0.961502     1\n",
      "1     cbx7    0.963570     2\n",
      "2     cbx8    0.976251     3\n",
      "3    ring1    0.979848     4\n",
      "4     brd7    0.980695     5\n",
      "5     hira    0.981533     6\n",
      "6    klf11    0.981574     7\n",
      "7      rb1    0.981926     8\n",
      "8    prkdc    0.982076     9\n",
      "9    sumo1    0.983449    10\n",
      "10    chd4    0.983453    11\n",
      "11   pcgf2    0.983623    12\n",
      "12    rbl1    0.983894    13\n",
      "13  ptpn11    0.984590    14\n",
      "14  anp32e    0.985184    15\n",
      "15   kdm6b    0.985256    16\n",
      "16    yap1    0.985587    17\n",
      "17    phf2    0.985973    18\n",
      "18    e2f7    0.986166    19\n",
      "19   esco2    0.986539    20\n",
      "20  polr2b    0.986764    21\n",
      "21  ints11    0.986792    22\n",
      "22    ubn1    0.986882    23\n",
      "23   fosl2    0.987221    24\n",
      "24    nfic    0.987277    25\n",
      "Finished stage 3 (exp)\n",
      "Stage 2 (acc): train ChromBERT to predict chromatin accessibility changes in cell state transition\n",
      "\n",
      "[Attempt 0/2] seed=55\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Epoch 0:  20%|████▍                 | 800/4000 [02:19<09:19,  5.71it/s, v_num=0]\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.15it/s]\u001b[A\n",
      "Epoch 0:  40%|▍| 1600/4000 [05:03<07:35,  5.27it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.15it/s]\u001b[A\n",
      "Epoch 0:  60%|▌| 2400/4000 [07:47<05:11,  5.13it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.01it/s]\u001b[A\n",
      "Epoch 0:  80%|▊| 3200/4000 [10:31<02:37,  5.07it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.00it/s]\u001b[A\n",
      "Epoch 0: 100%|█| 4000/4000 [13:15<00:00,  5.03it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.19it/s]\u001b[A\n",
      "Epoch 1:  20%|▏| 800/4000 [02:17<09:11,  5.80it/s, v_num=0, default_validation/r\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.17it/s]\u001b[A\n",
      "Epoch 1:  40%|▍| 1600/4000 [05:01<07:32,  5.31it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.17it/s]\u001b[A\n",
      "Epoch 1:  60%|▌| 2400/4000 [07:45<05:10,  5.16it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.14it/s]\u001b[A\n",
      "Epoch 1:  80%|▊| 3200/4000 [10:28<02:37,  5.09it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.14it/s]\u001b[A\n",
      "Epoch 1: 100%|█| 4000/4000 [13:12<00:00,  5.05it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.17it/s]\u001b[A\n",
      "Epoch 2:  20%|▏| 800/4000 [02:18<09:12,  5.79it/s, v_num=0, default_validation/r\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.14it/s]\u001b[A\n",
      "Epoch 2:  40%|▍| 1600/4000 [05:02<07:33,  5.29it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.15it/s]\u001b[A\n",
      "Epoch 2:  60%|▌| 2400/4000 [07:45<05:10,  5.16it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.13it/s]\u001b[A\n",
      "Epoch 2:  80%|▊| 3200/4000 [10:29<02:37,  5.09it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.18it/s]\u001b[A\n",
      "Epoch 2: 100%|█| 4000/4000 [13:12<00:00,  5.05it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.19it/s]\u001b[A\n",
      "Epoch 3:  20%|▏| 800/4000 [02:18<09:12,  5.79it/s, v_num=0, default_validation/r\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.17it/s]\u001b[A\n",
      "Epoch 3:  40%|▍| 1600/4000 [05:00<07:31,  5.32it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.18it/s]\u001b[A\n",
      "Epoch 3:  60%|▌| 2400/4000 [07:43<05:09,  5.18it/s, v_num=0, default_validation/\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation: |                                             | 0/? [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0:   0%|                          | 0/250 [00:00<?, ?it/s]\u001b[A\n",
      "Validation DataLoader 0: 100%|████████████████| 250/250 [00:24<00:00, 10.18it/s]\u001b[A\n",
      "Epoch 3:  60%|▌| 2400/4000 [08:08<05:25,  4.91it/s, v_num=0, default_validation/\u001b[A\n",
      "Evaluating the finetuned model performance\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/acc/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=2-step=163.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 110/110 parameters\n",
      "ft_ckpt: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/acc/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=2-step=163.ckpt, test_metrics: {'pearsonr': 0.8940250277519226, 'spearmanr': 0.8104968667030334, 'mse': 0.23963338136672974, 'mae': 0.344401478767395, 'r2': 0.7987537207411504}\n",
      "Attempt metrics: pearsonr=0.8940250277519226\n",
      "Accepted run (pearsonr=0.8940 >= 0.2).\n",
      "\n",
      "Finished stage 2: obtained a fine-tuned ChromBERT\n",
      "Best pearsonr=0.8940250277519226, metrics={'pearsonr': 0.8940250277519226, 'spearmanr': 0.8104968667030334, 'mse': 0.23963338136672974, 'mae': 0.344401478767395, 'r2': 0.7987537207411504, 'ft_ckpt': '/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/acc/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=2-step=163.ckpt'}\n",
      "Finished stage 2 (acc): train ChromBERT to predict chromatin accessibility changes in cell state transition\n",
      "Stage 3 (acc): infer driver factors in different chromatin accessibility activity genes\n",
      "        factors  similarity  rank\n",
      "0          myog    0.163047     1\n",
      "1   pax3-foxo1a    0.202741     2\n",
      "2          myf5    0.212604     3\n",
      "3         myod1    0.242216     4\n",
      "4          pax7    0.346181     5\n",
      "5          dux4    0.372867     6\n",
      "6          six2    0.376185     7\n",
      "7       neurog2    0.444524     8\n",
      "8         tead1    0.455782     9\n",
      "9          yap1    0.487734    10\n",
      "10      neurod1    0.492000    11\n",
      "11        pgbd3    0.495305    12\n",
      "12         six1    0.496939    13\n",
      "13         pax6    0.503089    14\n",
      "14        ascl1    0.508561    15\n",
      "15        tcf21    0.526704    16\n",
      "16       pou3f2    0.531527    17\n",
      "17         tbx5    0.545569    18\n",
      "18        znf92    0.579399    19\n",
      "19         lhx2    0.614269    20\n",
      "20         mycn    0.617738    21\n",
      "21         tead    0.668162    22\n",
      "22         ss18    0.669292    23\n",
      "23         chd4    0.700076    24\n",
      "24         otx2    0.705928    25\n",
      "Finished stage 4 acc: infer driver factors in cell state transition (top 25):\n",
      "        factors  similarity  rank\n",
      "0          myog    0.163047     1\n",
      "1   pax3-foxo1a    0.202741     2\n",
      "2          myf5    0.212604     3\n",
      "3         myod1    0.242216     4\n",
      "4          pax7    0.346181     5\n",
      "5          dux4    0.372867     6\n",
      "6          six2    0.376185     7\n",
      "7       neurog2    0.444524     8\n",
      "8         tead1    0.455782     9\n",
      "9          yap1    0.487734    10\n",
      "10      neurod1    0.492000    11\n",
      "11        pgbd3    0.495305    12\n",
      "12         six1    0.496939    13\n",
      "13         pax6    0.503089    14\n",
      "14        ascl1    0.508561    15\n",
      "15        tcf21    0.526704    16\n",
      "16       pou3f2    0.531527    17\n",
      "17         tbx5    0.545569    18\n",
      "18        znf92    0.579399    19\n",
      "19         lhx2    0.614269    20\n",
      "20         mycn    0.617738    21\n",
      "21         tead    0.668162    22\n",
      "22         ss18    0.669292    23\n",
      "23         chd4    0.700076    24\n",
      "24         otx2    0.705928    25\n",
      "Finished stage 3 (acc)\n",
      "Merging factor ranks from expression and chromatin accessibility\n"
     ]
    }
   ],
   "source": [
    "# Runtime estimates:\n",
    "#   - fast mode: ~3-5 hours\n",
    "#     (uses all ~19,620 genes for expression analysis, but downsamples \n",
    "#      chromatin accessibility regions to 20k for faster training)\n",
    "#\n",
    "# Note: Both modes (fast and full) use the complete gene expression dataset. The 'fast' mode \n",
    "# only downsamples chromatin accessibility regions, not gene data.\n",
    "\n",
    "# So this downsampled 5000 genes for expression analysis for test (40-100 minutes)\n",
    "\n",
    "!chrombert-tools find_driver_in_transition \\\n",
    "  --exp-tpm1 \"../data/fibroblast_expression_sample5000.csv\" \\\n",
    "  --exp-tpm2 \"../data/myoblast_expression_sample5000.csv\" \\\n",
    "  --acc-peak1 \"../data/fibroblast_ENCFF184KAM_peak.bed\" \\\n",
    "  --acc-peak2 \"../data/myoblast_ENCFF647RNC_peak.bed\" \\\n",
    "  --acc-signal1 \"../data/fibroblast_ENCFF361BTT_signal.bigwig\" \\\n",
    "  --acc-signal2 \"../data/myoblast_ENCFF149ERN_signal.bigwig\" \\\n",
    "  --genome 'hg38' \\\n",
    "  --resolution '1kb' \\\n",
    "  --odir output_find_driver_in_transition \\\n",
    "  --direction \"2-1\" 2> \"./tmp/hg38_1kb.stderr.log\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1278c59",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3fb4ee4f",
   "metadata": {},
   "source": [
    "### Load the fine-tuned checkpoint to infer key regulators and TRN for myoblast (skip fine-tuning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c168340e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage 1: prepare dataset\n",
      "Expression dataset already exists in output_find_driver_in_transition/exp/dataset\n",
      "Chromatin accessibility dataset already exists in output_find_driver_in_transition/acc/dataset\n",
      "Finished Stage 1\n",
      "Whether to train ChromBERT to predict expression changes in cell state transition: True\n",
      "Whether to train ChromBERT to predict chromatin accessibility changes in cell state transition: True\n",
      "Processing stage 2 (exp): train ChromBERT to predict expression changes in cell state transition\n",
      "Use fine-tuned ChromBERT checkpoint file: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/exp/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=0-step=21.ckpt to find driver factors in different expression activity genes\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/exp/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=0-step=21.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 110/110 parameters\n",
      "Finished Stage 2 (exp): use fine-tuned ChromBERT to find driver factors in different expression activity genes\n",
      "Stage 3 (exp): infer driver factors in different expression activity genes\n",
      "Finished stage 4 exp: infer driver factors in cell state transition (top 25):\n",
      "   factors  similarity  rank\n",
      "0     cbx6    0.961502     1\n",
      "1     cbx7    0.963570     2\n",
      "2     cbx8    0.976251     3\n",
      "3    ring1    0.979848     4\n",
      "4     brd7    0.980695     5\n",
      "5     hira    0.981533     6\n",
      "6    klf11    0.981574     7\n",
      "7      rb1    0.981926     8\n",
      "8    prkdc    0.982076     9\n",
      "9    sumo1    0.983449    10\n",
      "10    chd4    0.983453    11\n",
      "11   pcgf2    0.983623    12\n",
      "12    rbl1    0.983894    13\n",
      "13  ptpn11    0.984590    14\n",
      "14  anp32e    0.985184    15\n",
      "15   kdm6b    0.985256    16\n",
      "16    yap1    0.985587    17\n",
      "17    phf2    0.985973    18\n",
      "18    e2f7    0.986166    19\n",
      "19   esco2    0.986539    20\n",
      "20  polr2b    0.986764    21\n",
      "21  ints11    0.986792    22\n",
      "22    ubn1    0.986882    23\n",
      "23   fosl2    0.987221    24\n",
      "24    nfic    0.987277    25\n",
      "Finished stage 3 (exp)\n",
      "Use fine-tuned ChromBERT checkpoint file: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/acc/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=2-step=163.ckpt to find driver factors in different chromatin accessibility activity genes\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/acc/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=2-step=163.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 110/110 parameters\n",
      "Finished Stage 2 (acc): use fine-tuned ChromBERT to find driver factors in different chromatin accessibility activity genes\n",
      "Stage 3 (acc): infer driver factors in different chromatin accessibility activity genes\n",
      "Finished stage 4 acc: infer driver factors in cell state transition (top 25):\n",
      "        factors  similarity  rank\n",
      "0          myog    0.163047     1\n",
      "1   pax3-foxo1a    0.202741     2\n",
      "2          myf5    0.212604     3\n",
      "3         myod1    0.242216     4\n",
      "4          pax7    0.346181     5\n",
      "5          dux4    0.372867     6\n",
      "6          six2    0.376185     7\n",
      "7       neurog2    0.444524     8\n",
      "8         tead1    0.455782     9\n",
      "9          yap1    0.487734    10\n",
      "10      neurod1    0.492000    11\n",
      "11        pgbd3    0.495305    12\n",
      "12         six1    0.496939    13\n",
      "13         pax6    0.503089    14\n",
      "14        ascl1    0.508561    15\n",
      "15        tcf21    0.526704    16\n",
      "16       pou3f2    0.531527    17\n",
      "17         tbx5    0.545569    18\n",
      "18        znf92    0.579399    19\n",
      "19         lhx2    0.614269    20\n",
      "20         mycn    0.617738    21\n",
      "21         tead    0.668162    22\n",
      "22         ss18    0.669292    23\n",
      "23         chd4    0.700076    24\n",
      "24         otx2    0.705928    25\n",
      "Finished stage 3 (acc)\n",
      "Merging factor ranks from expression and chromatin accessibility\n",
      "Finished merging factor ranks from expression and chromatin accessibility\n",
      "Top 25 driver factors:\n",
      "    factors  similarity_exp  rank_exp  similarity_acc  rank_acc  total_rank\n",
      "0      yap1        0.985587        17        0.487734        10           1\n",
      "1      chd4        0.983453        11        0.700076        24           2\n",
      "2   neurog2        0.987556        29        0.444524         8           3\n",
      "3     tead1        0.988670        39        0.455782         9           4\n",
      "4     tcf21        0.987744        32        0.526704        16           4\n",
      "5     pgbd3        0.988605        38        0.495305        12           6\n",
      "6      myf5        0.989661        50        0.212604         3           7\n",
      "7       rb1        0.981926         8        0.828548        46           8\n",
      "8     myod1        0.990031        55        0.242216         4           9\n",
      "9      six1        0.989994        54        0.496939        13          10\n",
      "10    nr3c1        0.987676        31        0.796995        42          11\n",
      "11    fgfr1        0.989569        47        0.734220        30          12\n",
      "12     hira        0.981533         6        0.893640        71          12\n",
      "13     ubn1        0.986882        23        0.857598        57          14\n",
      "14    prmt5        0.987366        28        0.855554        56          15\n",
      "15    wwtr1        0.989648        49        0.778999        35          15\n",
      "16    klf11        0.981574         7        0.907471        81          17\n",
      "17    esco2        0.986539        20        0.892073        69          18\n",
      "18     tbx5        0.990667        73        0.545569        18          19\n",
      "19    foxo1        0.990244        61        0.796832        41          20\n",
      "20     rbl1        0.983894        13        0.913208        92          21\n",
      "21     chd7        0.990752        77        0.758251        32          22\n",
      "22    ring1        0.979848         4        0.923826       114          23\n",
      "23    sumo1        0.983449        10        0.922957       109          24\n",
      "24     cbx6        0.961502         1        0.925670       119          25\n",
      "Finished all stages!\n",
      "Used fine-tuned ChromBERT checkpoint: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/exp/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=0-step=21.ckpt\n",
      "Driver factors for expression changes: output_find_driver_in_transition/exp/results/factor_importance_rank.csv\n",
      "Used fine-tuned ChromBERT checkpoint: /mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/acc/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=2-step=163.ckpt\n",
      "Driver factors for chromatin accessibility changes: output_find_driver_in_transition/acc/results/factor_importance_rank.csv\n",
      "Integrated driver factor rankings: output_find_driver_in_transition/merge/factor_importance_rank.csv\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools find_driver_in_transition \\\n",
    "  --ft-ckpt-exp \"/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/exp/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=0-step=21.ckpt\" \\\n",
    "  --ft-ckpt-acc \"/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/acc/train/try_00_seed_55/lightning_logs/lightning_logs/version_0/checkpoints/epoch=2-step=163.ckpt\" \\\n",
    "  --genome 'hg38' \\\n",
    "  --exp-tpm1 \"../data/fibroblast_expression_sample5000.csv\" \\\n",
    "  --exp-tpm2 \"../data/myoblast_expression_sample5000.csv\" \\\n",
    "  --acc-peak1 \"../data/fibroblast_ENCFF184KAM_peak.bed\" \\\n",
    "  --acc-peak2 \"../data/myoblast_ENCFF647RNC_peak.bed\" \\\n",
    "  --acc-signal1 \"../data/fibroblast_ENCFF361BTT_signal.bigwig\" \\\n",
    "  --acc-signal2 \"../data/myoblast_ENCFF149ERN_signal.bigwig\" \\\n",
    "  --resolution '1kb' \\\n",
    "  --odir output_find_driver_in_transition \\\n",
    "  --direction \"2-1\" 2> \"./tmp/hg38_1kb.stderr.log\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "42705ea7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>factors</th>\n",
       "      <th>similarity</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cbx6</td>\n",
       "      <td>0.961502</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cbx7</td>\n",
       "      <td>0.963570</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cbx8</td>\n",
       "      <td>0.976251</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ring1</td>\n",
       "      <td>0.979848</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brd7</td>\n",
       "      <td>0.980695</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>stat4</td>\n",
       "      <td>0.996563</td>\n",
       "      <td>987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>gtf2i</td>\n",
       "      <td>0.996564</td>\n",
       "      <td>988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>irf5</td>\n",
       "      <td>0.996580</td>\n",
       "      <td>989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>tfcp2</td>\n",
       "      <td>0.996936</td>\n",
       "      <td>990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>mafg</td>\n",
       "      <td>0.997064</td>\n",
       "      <td>991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>991 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    factors  similarity  rank\n",
       "0      cbx6    0.961502     1\n",
       "1      cbx7    0.963570     2\n",
       "2      cbx8    0.976251     3\n",
       "3     ring1    0.979848     4\n",
       "4      brd7    0.980695     5\n",
       "..      ...         ...   ...\n",
       "986   stat4    0.996563   987\n",
       "987   gtf2i    0.996564   988\n",
       "988    irf5    0.996580   989\n",
       "989   tfcp2    0.996936   990\n",
       "990    mafg    0.997064   991\n",
       "\n",
       "[991 rows x 3 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "odir = \"./output_find_driver_in_transition\"\n",
    "exp_odir = f'{odir}/exp'\n",
    "exp_results_odir = f\"{exp_odir}/results\"\n",
    "exp_rank_df = pd.read_csv(os.path.join(exp_results_odir, \"factor_importance_rank.csv\"))\n",
    "exp_rank_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "67be6ce9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>factors</th>\n",
       "      <th>similarity</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>myog</td>\n",
       "      <td>0.163047</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pax3-foxo1a</td>\n",
       "      <td>0.202741</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>myf5</td>\n",
       "      <td>0.212604</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>myod1</td>\n",
       "      <td>0.242216</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pax7</td>\n",
       "      <td>0.346181</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>znf28</td>\n",
       "      <td>0.981777</td>\n",
       "      <td>987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>znf26</td>\n",
       "      <td>0.981883</td>\n",
       "      <td>988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>znf714</td>\n",
       "      <td>0.982148</td>\n",
       "      <td>989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>znf266</td>\n",
       "      <td>0.982677</td>\n",
       "      <td>990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>znf425</td>\n",
       "      <td>0.982688</td>\n",
       "      <td>991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>991 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         factors  similarity  rank\n",
       "0           myog    0.163047     1\n",
       "1    pax3-foxo1a    0.202741     2\n",
       "2           myf5    0.212604     3\n",
       "3          myod1    0.242216     4\n",
       "4           pax7    0.346181     5\n",
       "..           ...         ...   ...\n",
       "986        znf28    0.981777   987\n",
       "987        znf26    0.981883   988\n",
       "988       znf714    0.982148   989\n",
       "989       znf266    0.982677   990\n",
       "990       znf425    0.982688   991\n",
       "\n",
       "[991 rows x 3 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acc_odir = f'{odir}/acc'\n",
    "acc_results_odir = f\"{acc_odir}/results\"\n",
    "acc_rank_df = pd.read_csv(os.path.join(acc_results_odir, \"factor_importance_rank.csv\"))\n",
    "acc_rank_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "324a13a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>factors</th>\n",
       "      <th>similarity_exp</th>\n",
       "      <th>rank_exp</th>\n",
       "      <th>similarity_acc</th>\n",
       "      <th>rank_acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cbx6</td>\n",
       "      <td>0.961502</td>\n",
       "      <td>1</td>\n",
       "      <td>0.925670</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cbx7</td>\n",
       "      <td>0.963570</td>\n",
       "      <td>2</td>\n",
       "      <td>0.927905</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cbx8</td>\n",
       "      <td>0.976251</td>\n",
       "      <td>3</td>\n",
       "      <td>0.930284</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ring1</td>\n",
       "      <td>0.979848</td>\n",
       "      <td>4</td>\n",
       "      <td>0.923826</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brd7</td>\n",
       "      <td>0.980695</td>\n",
       "      <td>5</td>\n",
       "      <td>0.932991</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>stat4</td>\n",
       "      <td>0.996563</td>\n",
       "      <td>987</td>\n",
       "      <td>0.961282</td>\n",
       "      <td>519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>gtf2i</td>\n",
       "      <td>0.996564</td>\n",
       "      <td>988</td>\n",
       "      <td>0.976028</td>\n",
       "      <td>911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>irf5</td>\n",
       "      <td>0.996580</td>\n",
       "      <td>989</td>\n",
       "      <td>0.970488</td>\n",
       "      <td>787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>tfcp2</td>\n",
       "      <td>0.996936</td>\n",
       "      <td>990</td>\n",
       "      <td>0.980216</td>\n",
       "      <td>974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>mafg</td>\n",
       "      <td>0.997064</td>\n",
       "      <td>991</td>\n",
       "      <td>0.977517</td>\n",
       "      <td>934</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>991 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    factors  similarity_exp  rank_exp  similarity_acc  rank_acc\n",
       "0      cbx6        0.961502         1        0.925670       119\n",
       "1      cbx7        0.963570         2        0.927905       124\n",
       "2      cbx8        0.976251         3        0.930284       133\n",
       "3     ring1        0.979848         4        0.923826       114\n",
       "4      brd7        0.980695         5        0.932991       142\n",
       "..      ...             ...       ...             ...       ...\n",
       "986   stat4        0.996563       987        0.961282       519\n",
       "987   gtf2i        0.996564       988        0.976028       911\n",
       "988    irf5        0.996580       989        0.970488       787\n",
       "989   tfcp2        0.996936       990        0.980216       974\n",
       "990    mafg        0.997064       991        0.977517       934\n",
       "\n",
       "[991 rows x 5 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_df = pd.merge(exp_rank_df,acc_rank_df,on='factors',how='inner',suffixes=['_exp','_acc'])\n",
    "merge_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "eb30fa4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>factors</th>\n",
       "      <th>similarity_exp</th>\n",
       "      <th>rank_exp</th>\n",
       "      <th>similarity_acc</th>\n",
       "      <th>rank_acc</th>\n",
       "      <th>total_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>yap1</td>\n",
       "      <td>0.985587</td>\n",
       "      <td>17</td>\n",
       "      <td>0.487734</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chd4</td>\n",
       "      <td>0.983453</td>\n",
       "      <td>11</td>\n",
       "      <td>0.700076</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>neurog2</td>\n",
       "      <td>0.987556</td>\n",
       "      <td>29</td>\n",
       "      <td>0.444524</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tead1</td>\n",
       "      <td>0.988670</td>\n",
       "      <td>39</td>\n",
       "      <td>0.455782</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tcf21</td>\n",
       "      <td>0.987744</td>\n",
       "      <td>32</td>\n",
       "      <td>0.526704</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>znf266</td>\n",
       "      <td>0.996125</td>\n",
       "      <td>963</td>\n",
       "      <td>0.982677</td>\n",
       "      <td>990</td>\n",
       "      <td>986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>maff</td>\n",
       "      <td>0.996367</td>\n",
       "      <td>977</td>\n",
       "      <td>0.980712</td>\n",
       "      <td>979</td>\n",
       "      <td>988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>znf26</td>\n",
       "      <td>0.996320</td>\n",
       "      <td>974</td>\n",
       "      <td>0.981883</td>\n",
       "      <td>988</td>\n",
       "      <td>989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>pitx1</td>\n",
       "      <td>0.996537</td>\n",
       "      <td>986</td>\n",
       "      <td>0.980594</td>\n",
       "      <td>977</td>\n",
       "      <td>990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>tfcp2</td>\n",
       "      <td>0.996936</td>\n",
       "      <td>990</td>\n",
       "      <td>0.980216</td>\n",
       "      <td>974</td>\n",
       "      <td>991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>991 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     factors  similarity_exp  rank_exp  similarity_acc  rank_acc  total_rank\n",
       "0       yap1        0.985587        17        0.487734        10           1\n",
       "1       chd4        0.983453        11        0.700076        24           2\n",
       "2    neurog2        0.987556        29        0.444524         8           3\n",
       "3      tead1        0.988670        39        0.455782         9           4\n",
       "4      tcf21        0.987744        32        0.526704        16           4\n",
       "..       ...             ...       ...             ...       ...         ...\n",
       "986   znf266        0.996125       963        0.982677       990         986\n",
       "987     maff        0.996367       977        0.980712       979         988\n",
       "988    znf26        0.996320       974        0.981883       988         989\n",
       "989    pitx1        0.996537       986        0.980594       977         990\n",
       "990    tfcp2        0.996936       990        0.980216       974         991\n",
       "\n",
       "[991 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_df['total_rank']=((merge_df['rank_exp']+merge_df['rank_acc'])/2).rank().astype(int)\n",
    "merge_df = merge_df.sort_values('total_rank').reset_index(drop=True)\n",
    "merge_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6d2c38fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>factors</th>\n",
       "      <th>similarity_exp</th>\n",
       "      <th>rank_exp</th>\n",
       "      <th>similarity_acc</th>\n",
       "      <th>rank_acc</th>\n",
       "      <th>total_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>myod1</td>\n",
       "      <td>0.990031</td>\n",
       "      <td>55</td>\n",
       "      <td>0.242216</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  factors  similarity_exp  rank_exp  similarity_acc  rank_acc  total_rank\n",
       "8   myod1        0.990031        55        0.242216         4           9"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_df.query(\"factors == 'myod1'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fa1c0546",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "merge_df = pd.read_csv(\"/mnt/Storage2/home/chenqianqian/projects/chrombert/chrombert_tools/ChromBERT-tools/examples/cli/output_find_driver_in_transition/merge/factor_importance_rank.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ca3e51a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>factors</th>\n",
       "      <th>similarity_exp</th>\n",
       "      <th>rank_exp</th>\n",
       "      <th>similarity_acc</th>\n",
       "      <th>rank_acc</th>\n",
       "      <th>total_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>yap1</td>\n",
       "      <td>0.985587</td>\n",
       "      <td>17</td>\n",
       "      <td>0.487734</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chd4</td>\n",
       "      <td>0.983453</td>\n",
       "      <td>11</td>\n",
       "      <td>0.700076</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>neurog2</td>\n",
       "      <td>0.987556</td>\n",
       "      <td>29</td>\n",
       "      <td>0.444524</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tead1</td>\n",
       "      <td>0.988670</td>\n",
       "      <td>39</td>\n",
       "      <td>0.455782</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tcf21</td>\n",
       "      <td>0.987744</td>\n",
       "      <td>32</td>\n",
       "      <td>0.526704</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   factors  similarity_exp  rank_exp  similarity_acc  rank_acc  total_rank\n",
       "0     yap1        0.985587        17        0.487734        10           1\n",
       "1     chd4        0.983453        11        0.700076        24           2\n",
       "2  neurog2        0.987556        29        0.444524         8           3\n",
       "3    tead1        0.988670        39        0.455782         9           4\n",
       "4    tcf21        0.987744        32        0.526704        16           4"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_df.head(n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3555f9e7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "chrombert",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
