{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5b24070",
   "metadata": {},
   "source": [
    "# Impute cistromes\n",
    "\n",
    "The ``impute_cistrome`` command uses ChromBERT's learned co-association patterns to impute cistrome signals (e.g., ChIP-seq) for factor–cell pairs where experimental data is unavailable."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b47fea2",
   "metadata": {},
   "source": [
    "**Note**: The remaining examples will only show the direct command usage. \n",
    "\n",
    "If you need to use Singularity container, please refer to the [`singularity_use.ipynb`](singularity_use.ipynb) tutorial for detailed instructions on using `singularity exec` with `chrombert-tools`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08eab94f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "workdir=\"/mnt/Storage2/home/chenqianqian/projects/chrombert_tools/2.test/pull/ChromBERT-tools/examples/cli\" # your workdir\n",
    "os.chdir(workdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4f19688f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools [OPTIONS] COMMAND [ARGS]...\n",
      "\n",
      "  Type -h or --help after any subcommand for more information.\n",
      "\n",
      "Options:\n",
      "  -v, --verbose  Verbose logging\n",
      "  -d, --debug    Post mortem debugging\n",
      "  -V, --version  Show the version and exit.\n",
      "  -h, --help     Show this message and exit.\n",
      "\n",
      "Commands:\n",
      "  embed_cell_cistrome             Extract cell-specific cistrome...\n",
      "  embed_cell_gene                 Extract cell-specific gene embeddings\n",
      "  embed_cell_region               Extract cell-specific region embeddings\n",
      "  embed_cell_regulator            Extract cell-specific regulator...\n",
      "  embed_cistrome                  Extract general cistrome embeddings on...\n",
      "  embed_gene                      Extract general gene embeddings\n",
      "  embed_region                    Extract general region embeddings\n",
      "  embed_regulator                 Extract general regulator embeddings on...\n",
      "  find_context_specific_cofactor  Find context-specific cofactors in...\n",
      "  find_driver_in_transition       Find driver factors in cell state...\n",
      "  impute_cistrome                 Impute cistromes on specified regions\n",
      "  infer_cell_key_regulator        Infer cell-specific key regulators\n",
      "  infer_ep                        Infer enhancer-promoter loop\n",
      "  infer_regulator_network         Infer general regulator-regulator network\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f61703f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: chrombert-tools impute_cistrome [OPTIONS]\n",
      "\n",
      "  Impute cistromes on specified regions\n",
      "\n",
      "Options:\n",
      "  --region FILE                   Region BED file.  [required]\n",
      "  --cistrome TEXT                 factor:cell e.g.\n",
      "                                  BCL11A:GM12878;BRD4:MCF7;CTCF:HepG2. Use ';'\n",
      "                                  to separate multiple cistromes.  [required]\n",
      "  --odir DIRECTORY                Output directory.  [default: ./output]\n",
      "  --genome [hg38|mm10]            Genome.  [default: hg38]\n",
      "  --resolution [1kb]              Resolution. Only supports 1kb resolution in\n",
      "                                  imputing cistromes task.  [default: 1kb]\n",
      "  --batch-size INTEGER            Batch size. if you have enough GPU memory,\n",
      "                                  you can set it to a larger value.  [default:\n",
      "                                  4]\n",
      "  --chrombert-cache-dir DIRECTORY\n",
      "                                  ChromBERT cache directory (containing\n",
      "                                  config/ and anno/ subfolders).  [default:\n",
      "                                  ~/.cache/chrombert/data]\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools impute_cistrome -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6e61a57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region summary - total: 100, overlapping with ChromBERT: 100 (one region may overlap multiple ChromBERT regions, We keep overlaps with ≥50% coverage of either the ChromBERT bin or the input region),non-overlapping: 0\n",
      "celltype: h1 has no corresponding wild type dnase data in ChromBERT.\n",
      "Note: All cistromes names were converted to lowercase for matching.\n",
      "Cistromes count summary - requested: 6, matched in ChromBERT: 5, not found: 1, not found cistromes: ['myc:h1']\n",
      "ChromBERT cistromes metas: /mnt/Storage/home/chenqianqian/.cache/chrombert/data/config/hg38_6k_meta.tsv\n",
      "Your supervised_file does not contain the 'label' column. Please verify whether ground truth column ('label') is required. If it is not needed, you may disregard this message.\n",
      "Your supervised_file does not contain the 'label' column. Please verify whether ground truth column ('label') is required. If it is not needed, you may disregard this message.\n",
      "use organisim hg38; max sequence length is 6391\n",
      "use organisim hg38; max sequence length is 6391\n",
      "Loading checkpoint from /mnt/Storage/home/chenqianqian/.cache/chrombert/data/checkpoint/hg38_6k_1kb_prompt_cistrome.ckpt\n",
      "Loading from pl module, remove prefix 'model.'\n",
      "Loaded 112/112 parameters\n",
      "100%|███████████████████████████████████████████| 25/25 [00:04<00:00,  5.22it/s]\n",
      "Finished imputing cistromes on specific regions.\n",
      "Results saved to ./output_impute/results_prob_df.csv\n",
      "Results track files saved to ./output_impute/*.bw\n"
     ]
    }
   ],
   "source": [
    "!chrombert-tools impute_cistrome \\\n",
    "    --cistrome \"BCL11A:GM12878;BRD4:MCF7;CTCF:HepG2;MYC:H1;MYC:h9;SPI1:GSM2702714\" \\\n",
    "    --region \"../data/CTCF_ENCFF664UGR_sample100.bed\" \\\n",
    "    --odir \"./output_impute\" \\\n",
    "    --genome \"hg38\" \\\n",
    "    --resolution \"1kb\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7b508c52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input_chrom</th>\n",
       "      <th>input_start</th>\n",
       "      <th>input_end</th>\n",
       "      <th>chrombert_build_region_index</th>\n",
       "      <th>chrombert_start</th>\n",
       "      <th>chrombert_end</th>\n",
       "      <th>bcl11a:gm12878</th>\n",
       "      <th>brd4:mcf7</th>\n",
       "      <th>ctcf:hepg2</th>\n",
       "      <th>myc:h9</th>\n",
       "      <th>spi1:gsm2702714</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>37989946</td>\n",
       "      <td>37990368</td>\n",
       "      <td>32658</td>\n",
       "      <td>37990000</td>\n",
       "      <td>37991000</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>0.660156</td>\n",
       "      <td>0.984375</td>\n",
       "      <td>0.972656</td>\n",
       "      <td>0.632812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2400199</td>\n",
       "      <td>2400617</td>\n",
       "      <td>289179</td>\n",
       "      <td>2400000</td>\n",
       "      <td>2401000</td>\n",
       "      <td>0.664062</td>\n",
       "      <td>0.570312</td>\n",
       "      <td>0.972656</td>\n",
       "      <td>0.882812</td>\n",
       "      <td>0.917969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>6778809</td>\n",
       "      <td>6779319</td>\n",
       "      <td>391108</td>\n",
       "      <td>6779000</td>\n",
       "      <td>6780000</td>\n",
       "      <td>0.527344</td>\n",
       "      <td>0.412109</td>\n",
       "      <td>0.980469</td>\n",
       "      <td>0.871094</td>\n",
       "      <td>0.503906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>52980788</td>\n",
       "      <td>52981316</td>\n",
       "      <td>424926</td>\n",
       "      <td>52981000</td>\n",
       "      <td>52982000</td>\n",
       "      <td>0.174805</td>\n",
       "      <td>0.601562</td>\n",
       "      <td>0.976562</td>\n",
       "      <td>0.812500</td>\n",
       "      <td>0.345703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12</td>\n",
       "      <td>53676021</td>\n",
       "      <td>53676448</td>\n",
       "      <td>425578</td>\n",
       "      <td>53676000</td>\n",
       "      <td>53677000</td>\n",
       "      <td>0.494141</td>\n",
       "      <td>0.699219</td>\n",
       "      <td>0.968750</td>\n",
       "      <td>0.945312</td>\n",
       "      <td>0.570312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>chr6</td>\n",
       "      <td>53171843</td>\n",
       "      <td>53172315</td>\n",
       "      <td>1660979</td>\n",
       "      <td>53172000</td>\n",
       "      <td>53173000</td>\n",
       "      <td>0.408203</td>\n",
       "      <td>0.474609</td>\n",
       "      <td>0.988281</td>\n",
       "      <td>0.566406</td>\n",
       "      <td>0.617188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>chr6</td>\n",
       "      <td>131628105</td>\n",
       "      <td>131628616</td>\n",
       "      <td>1713078</td>\n",
       "      <td>131628000</td>\n",
       "      <td>131629000</td>\n",
       "      <td>0.632812</td>\n",
       "      <td>0.667969</td>\n",
       "      <td>0.988281</td>\n",
       "      <td>0.894531</td>\n",
       "      <td>0.773438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158704189</td>\n",
       "      <td>158704642</td>\n",
       "      <td>1735665</td>\n",
       "      <td>158704000</td>\n",
       "      <td>158705000</td>\n",
       "      <td>0.558594</td>\n",
       "      <td>0.251953</td>\n",
       "      <td>0.972656</td>\n",
       "      <td>0.613281</td>\n",
       "      <td>0.554688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>chr9</td>\n",
       "      <td>128117589</td>\n",
       "      <td>128118035</td>\n",
       "      <td>2049996</td>\n",
       "      <td>128117000</td>\n",
       "      <td>128118000</td>\n",
       "      <td>0.597656</td>\n",
       "      <td>0.468750</td>\n",
       "      <td>0.972656</td>\n",
       "      <td>0.812500</td>\n",
       "      <td>0.468750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>chr9</td>\n",
       "      <td>136122853</td>\n",
       "      <td>136123320</td>\n",
       "      <td>2057396</td>\n",
       "      <td>136123000</td>\n",
       "      <td>136124000</td>\n",
       "      <td>0.167969</td>\n",
       "      <td>0.310547</td>\n",
       "      <td>0.968750</td>\n",
       "      <td>0.394531</td>\n",
       "      <td>0.400391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   input_chrom  input_start  input_end  chrombert_build_region_index  \\\n",
       "0         chr1     37989946   37990368                         32658   \n",
       "1        chr11      2400199    2400617                        289179   \n",
       "2        chr12      6778809    6779319                        391108   \n",
       "3        chr12     52980788   52981316                        424926   \n",
       "4        chr12     53676021   53676448                        425578   \n",
       "..         ...          ...        ...                           ...   \n",
       "95        chr6     53171843   53172315                       1660979   \n",
       "96        chr6    131628105  131628616                       1713078   \n",
       "97        chr6    158704189  158704642                       1735665   \n",
       "98        chr9    128117589  128118035                       2049996   \n",
       "99        chr9    136122853  136123320                       2057396   \n",
       "\n",
       "    chrombert_start  chrombert_end  bcl11a:gm12878  brd4:mcf7  ctcf:hepg2  \\\n",
       "0          37990000       37991000        0.781250   0.660156    0.984375   \n",
       "1           2400000        2401000        0.664062   0.570312    0.972656   \n",
       "2           6779000        6780000        0.527344   0.412109    0.980469   \n",
       "3          52981000       52982000        0.174805   0.601562    0.976562   \n",
       "4          53676000       53677000        0.494141   0.699219    0.968750   \n",
       "..              ...            ...             ...        ...         ...   \n",
       "95         53172000       53173000        0.408203   0.474609    0.988281   \n",
       "96        131628000      131629000        0.632812   0.667969    0.988281   \n",
       "97        158704000      158705000        0.558594   0.251953    0.972656   \n",
       "98        128117000      128118000        0.597656   0.468750    0.972656   \n",
       "99        136123000      136124000        0.167969   0.310547    0.968750   \n",
       "\n",
       "      myc:h9  spi1:gsm2702714  \n",
       "0   0.972656         0.632812  \n",
       "1   0.882812         0.917969  \n",
       "2   0.871094         0.503906  \n",
       "3   0.812500         0.345703  \n",
       "4   0.945312         0.570312  \n",
       "..       ...              ...  \n",
       "95  0.566406         0.617188  \n",
       "96  0.894531         0.773438  \n",
       "97  0.613281         0.554688  \n",
       "98  0.812500         0.468750  \n",
       "99  0.394531         0.400391  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# results_pro_df: Imputed peak probabilities.\n",
    "results_pro_df = pd.read_csv(\"./output_impute/results_prob_df.csv\")\n",
    "results_pro_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33a477c6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37bc63a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be7e9841-ea1c-437e-b725-311cf57bcb8e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
